src/runtime_src/core/edge/skd/xrt_skd.cpp

```diff
@@ -0,0 +1,377 @@
+	//	    break;
+	//	  }
+	  ffi_arg_values[i] = &xrtHandle;
+	} else if(args[i].type == xrt_core::xclbin::kernel_argument::argtype::global) {
+	  uint64_t *buf_addr_ptr = (uint64_t*)(&args_from_host[args[i].offset/4]);
+	  uint64_t buf_addr = reinterpret_cast<uint64_t>(*buf_addr_ptr);
+	  uint64_t *buf_size_ptr = (uint64_t*)(&args_from_host[args[i].offset/4+2]);
+	  uint64_t buf_size = reinterpret_cast<uint64_t>(*buf_size_ptr);
+	  boSize[i] = buf_size;
+	  
+	  boHandles[i] = xclGetHostBO(devHdl,buf_addr,buf_size);
+	  bos[i] = xclMapBO(devHdl,boHandles[i],true);
+	  ffi_arg_values[i] = &bos[i];
+	  bo_list.emplace_back(i);
+	} else {
+	  ffi_arg_values[i] = &args_from_host[args[i].offset/4];
+	}
+      }
+      
+      ffi_call(&cif,FFI_FN(kernel), &kernel_return, ffi_arg_values);
+      args_from_host[1] = (uint32_t)kernel_return;
+      
+      // Unmap Buffers
+      for(auto i:bo_list) {
+	munmap(bos[i],boSize[i]);
+	xclFreeBO(devHdl,boHandles[i]);
+      }
+    }
+
+  }
+  
+  skd::~skd() {
+    // Call soft kernel fini if available
+    kernel_fini_t kernel_fini;
+    char sk_fini[PS_KERNEL_NAME_LENGTH+5];
+    int ret = 0;
+    snprintf(sk_fini,PS_KERNEL_NAME_LENGTH+5,"%s%s",sk_name,"_fini");
+    kernel_fini = (kernel_fini_t)dlsym(sk_handle, sk_fini);
+    if (!kernel_fini) {
+      syslog(LOG_INFO, "kernel fini function %s not found\n", sk_fini);
+    } else {
+      ret = kernel_fini(xrtHandle);
+    }
+    
+    dlclose(sk_handle);
+    ret = deleteSoftKernelFile();
+    if (ret) {
+      syslog(LOG_ERR, "Cannot remove soft kernel file %s\n", sk_path);
+    }
+    if(cmd_boh >= 0) {
+      xclBOProperties prop;
+      if (xclGetBOProperties(devHdl, cmd_boh, &prop)) {
+      }
+      ret = munmap(args_from_host,prop.size);
+      if (ret) {
+	syslog(LOG_ERR, "Cannot munmap BO %d, at %p\n", cmd_boh, &args_from_host);
+      }
+    }
+    xclClose(devHdl);
+  }
+
+  int skd::createSoftKernel(int *boh) {
+    int ret;
+    ret = xclSKCreate(devHdl, boh, cu_idx);
+    return ret;
+  }
+  int skd::waitNextCmd() {
+    return xclSKReport(devHdl, cu_idx, XRT_SCU_STATE_DONE);
+  }
+
+  /*
+   * This function create a soft kernel file.
+   */
+  int skd::createSoftKernelFile(xclDeviceHandle handle, int bohdl)
+  {
+    FILE *fptr;
+    void *buf;
+    char path[XRT_MAX_PATH_LENGTH];
+    int len, i;
+    
+    xclBOProperties prop;
+    if (xclGetBOProperties(handle, bohdl, &prop)) {
+      syslog(LOG_ERR, "Cannot get SK .so BO info.\n");
+      xclFreeBO(handle, bohdl);
+      return -1;
+    }
+    
+    buf = xclMapBO(handle, bohdl, false);
+    if (!buf) {
+      syslog(LOG_ERR, "Cannot map softkernel BO.\n");
+      xclFreeBO(handle, bohdl);
+      return -1;
+    }
+    
+    snprintf(path, XRT_MAX_PATH_LENGTH, "%s", SOFT_KERNEL_FILE_PATH);
+    len = strlen(path);
+    
+    /* If not exist, create the path one by one. */
+    for (i = 1; i < len; i++) {
```

Comment:
This should be a helper function or std::filesystem::create_directories should be used.